src/utp_stream.cpp

```diff
@@ -1813,22 +1823,38 @@ bool utp_socket_impl::send_pkt(int const flags)
 		&& m_write_buffer_size >= m_mtu_floor * 3
 		&& m_seq_nr != 0
 		&& (m_cwnd >> 16) > m_mtu_floor * 3);
+	// for non MTU-probes, use the conservative packet size
+	int const effective_mtu = mtu_probe ? m_mtu : m_mtu_floor;
+
+	boost::uint32_t const close_reason = m_close_reason;
+
+
+	int sack = 0;
+	if (m_inbuf.size())
+	{
+        const int max_sack_size = effective_mtu 
+            - sizeof(utp_header)
+            - 2 // for sack padding/header
+            - (close_reason ? 6 : 0);
+
+		// the SACK bitfield should ideally fit all
+		// the pieces we have successfully received
+		sack = (m_inbuf.span() + 7) / 8;
+		if (sack > max_sack_size) sack = max_sack_size;
```

Comment:
is this really sufficient? I seem to recall the code that actually generates the SACK header also to be constrained to a unt32

Reply:
Yes, this is really sufficient. `write_sack` uses `sack` variable, that denotes the SACK size in bytes. It writes the size of SACK in the packet itself. That is where `parse_sack` and `remove_sack_header` use that.
```
void utp_socket_impl::write_sack(boost::uint8_t* buf, int size) const
{
    INVARIANT_CHECK;

    TORRENT_ASSERT(m_inbuf.size());
    int ack_nr = (m_ack_nr + 2) & ACK_MASK;
    boost::uint8_t* end = buf + size;

    for (; buf != end; ++buf)
    {
        *buf = 0;
        int mask = 1;
        for (int i = 0; i < 8; ++i)
        {
            if (m_inbuf.at(ack_nr)) *buf |= mask;
            mask <<= 1;
            ack_nr = (ack_nr + 1) & ACK_MASK;
        }
    }
}
```

Anyway, it would be nice to add huge SACK test to unit tests.

Reply:
yeah, I took some notes. the ``parse_sack()``, ``write_sack()`` and ``remove_sack()`` could most likely be made free functions and easily unit tested. I'm mainly working on a bigger issue in RC_1_1 right now though.

Reply:
When I solve the variance problem and other more important issues with this patch, I could take this problem, and write unit tests.